Resolving workflows for REST requests

slobodanadamovic opened the pull request:
This PR adds ability to resolve a workflow based on a REST handler 
that handles a user's HTTP request. The resolved workflow is set in
a thread context and will be used to enforce workflows restriction 
defined on the API key.

This PR only focuses on definition and resolution of workflows.
Enforcing of workflows for API keys will be handled in a followup PRs.

Relates to https://github.com/elastic/elasticsearch/pull/96215

slobodanadamovic committed:
Define workflow and workflow resolver

slobodanadamovic committed:
Test workflow resolution

slobodanadamovic committed:
Fix assertion

slobodanadamovic committed:
Apply spotless

slobodanadamovic committed:
Renaming

slobodanadamovic committed:
Resolve workflow and store in thread context

slobodanadamovic committed:
Test security filter + small nit refactoring

elasticsearchmachine commented:
Pinging @elastic/es-security (Team:Security)

slobodanadamovic commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/restriction/WorkflowResolver.java:
```diff
@@ -0,0 +1,157 @@
+    /**
+     * Allows access to Search Application query REST API.
+     */
+    public static final Workflow SEARCH_APPLICATION_QUERY_WORKFLOW = Workflow.builder()
```
The `SEARCH_APPLICATION_QUERY_WORKFLOW` and `SEARCH_APPLICATION_ANALYTICS_WORKFLOW` are just a suggestion that might need to be changed depending on the result of an internal discussion.

slobodanadamovic replied:
We synced internally and agreed that (for now) we only need a single workflow `search_application_query` that restricts access to `RestQuerySearchApplicationAction`. I've adjusted the workflows accordingly.

ywangd commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/restriction/WorkflowResolver.java:
```diff
@@ -0,0 +1,157 @@
+     */
+    public static final Workflow SEARCH_APPLICATION_QUERY_WORKFLOW = Workflow.builder()
+        .name("search_application_query")
+        .addAllowedRestHandlers("search_application_query_action")
```
If `RestQuerySearchApplicationAction` changes its name, the value here won't be automatically updated. An alternative is to let each RestHandler declare its (optional) workflow name in addition to the existing `name`. But that would bind workflow tightly to RestHandler and does not seem ideal. Maybe we can just add relevant tests in future PRs to ensure the names here are in sync with the RestHandlers.

slobodanadamovic replied:
> Maybe we can just add relevant tests in future PRs to ensure the names here are in sync with the RestHandlers.

As discussed in sync, this is a valid concern and relying on having a relevant tests per workflow should be sufficient to detect name changes.

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/restriction/WorkflowService.java:
```diff
@@ -0,0 +1,78 @@
+
+    public static final String WORKFLOW_HEADER = "_xpack_security_workflow";
+
+    private final XPackLicenseState licenseState;
```
Now I think we might not need license check at all. The search application APIs already [enforce license checks](https://github.com/elastic/elasticsearch/blob/1cc1d124327fca6a638aac8656873336b64fee5a/x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/action/SearchApplicationTransportAction.java#L55). Unless we want to license the workflow feature separately (and I don't think that's the case), we can just proceed regardlessly here.

PS: The current license check for enterprise search APIs is done at transport layer. I think it's better to be moved earlier to REST layer. I have raised a question in the slack channel.

slobodanadamovic replied:
Thanks for raising this. After internal discussions, we agreed that it is not necessary to add a license check here since workflows are simply allowing existing user-facing feature (Search Application) to work correctly. The Search Application has already implemented proper license checks, which make license check of workflows unnecessary and redundant.

ywangd reviewed (commented):
I haven't read the test code yet. Here are some comments for the main code.

slobodanadamovic committed:
Refactorings:

- Remove unused license state
- Remove analytics workflow
- Remove action listener from resolveWorkflowAndStoreInThreadContext
- Rename query workflow
- Remove reflection

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-workflows-resolution

slobodanadamovic commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/restriction/WorkflowService.java:
```diff
@@ -0,0 +1,78 @@
+        if (restHandler instanceof BaseRestHandler baseRestHandler) {
+            return baseRestHandler.getName();
+        } else if (restHandler.getConcreteRestHandler() instanceof BaseRestHandler baseRestHandler) {
+            return baseRestHandler.getName();
```
We agreed that it is not necessary to check `RestHandler.getConcreteRestHandler` and that we should consider introducing `getName()` to the `RestHandler` interface in a followup PR.

slobodanadamovic committed:
Encapsulate access to workflow header

slobodanadamovic committed:
Fix assertion

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-workflows-resolution

ywangd commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/restriction/WorkflowResolver.java:
```diff
@@ -0,0 +1,157 @@
+     * @throws IllegalArgumentException if a workflow with the given {@code name} does not exist
+     */
+    public static Workflow resolveWorkflowByName(final String name) {
+        final String filteredName = Objects.requireNonNull(name).toLowerCase(Locale.ROOT).trim();
```
I think this depends on whether we consider workflow more like a privilege or more like a resource. Privilege names are often case-insensitive as you stated. A resource is always case-sensitive, e.g. index patterns, run-as username. In my mind, a workflow is more like resource, i.e. it is a group of APIs for specific product purpose. So I'd prefer case sensitive. But I'll leave it to you for decision.

slobodanadamovic replied:
Honestly, I haven't thought about a workflow as resource. I've been treating workflows as privileges that allow executing certain REST actions. But I can see your point and agree it feels more like a resource.

ywangd commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/rest/SecurityRestFilterTests.java:
```diff
@@ -258,5 +280,55 @@ public void testSanitizeHeaders() throws Exception {
+        if (randomBoolean()) {
+            restHandler = new TestBaseRestHandler(randomFrom(workflow.allowedRestHandlers()));
+        } else {
+            restHandler = Mockito.mock(RestHandler.class);
+            when(restHandler.getConcreteRestHandler()).thenReturn(new TestBaseRestHandler(randomFrom(workflow.allowedRestHandlers())));
```
I think this does not work anymore since we agreed to not check the wrapped restHandler?

slobodanadamovic replied:
Good catch. Removed it. 👍

ywangd reviewed (approved):
LGTM

You didn't re-request my review. But I took the liberty to assume that the changes are ready. Thanks!

slobodanadamovic committed:
Remove usage of restHandler.getConcreteRestHandler in tests

slobodanadamovic committed:
Assert allowed REST handlers are not empty

slobodanadamovic committed:
Make workflow names case insensitive

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-workflows-resolution
